File: Trees, Binary Trees, & Binary Search Trees/HeightBalancedBinaryTree/HeightBalancedBinaryTree.py
"""
    This solution taken from a book "Elements of Programming Interviews".
    Some variable names has been changed and code adapted  the solution to pass Leetocde
    Balanced Binary Tree - LeetCode: https://leetcode.com/problems/balanced-binary-tree

    Common misunderstanding with this questions is that canditates understood that balance can be
    broken on root. But as said in task conditions:
    "a binary tree in which the left and right subtrees of every node differ in height by no more than 1."
    We must check balance for each node.
    Also we care about 2 things as our recursion goes upwards after bottoming out:
        - The node's height
        -  Whether its left and right subtrees are balanced

    The video to explain this code is here: https://www.youtube.com/watch?v=LU4fGD-fgJQ

"""


import collections


class Solution(object):
    def isBalanced(self, root):
        BalanedStatusWithHeight = collections.namedtuple('BalanedStatusWithHeight', ('balanced', 'height'))

        # First value of the return value indicate if tree is balanced and if
        # balanced teh second value of teh return value is the height of tree.
        def check_balanced(root):
            if not root:
                return BalanedStatusWithHeight(True, -1)  # Base case

            left_result = check_balanced(root.left)
            if not left_result.balanced:
                # Left subtree is not balanced
                return BalanedStatusWithHeight(False, 0)

            right_result = check_balanced(root.right)
            if not right_result.balanced:
                # Right subtree is not balanced
                return BalanedStatusWithHeight(False, 0)

            is_balanced = abs(left_result.height - right_result.height) <= 1
            height = max(left_result.height, right_result.height) + 1
            return BalanedStatusWithHeight(is_balanced, height)

        return check_balanced(root).balanced

"""
If we replace namedtuple with simple array list the code will work 172/36 = 4.7 times more speedy
"""


class Solution(object):
    def isBalanced(self, root):
        def checkBalanced(root):

            if not root:
                return [True, -1]

            # check left subtree
            left_result = checkBalanced(root.left)
            if not left_result[0]:
                return [False, 0]

            # check right subtree
            right_result = checkBalanced(root.right)
            if not right_result[0]:
                return [False, 0]

            is_balanced = abs(left_result[1] - right_result[1]) <= 1
            height = max(left_result[1], right_result[1]) + 1
            return [is_balanced, height]

        return checkBalanced(root)[0]

File: Trees, Binary Trees, & Binary Search Trees/HeightBalancedBinaryTree/HeightBalancedBinaryTree.java
/*
  This repository is no longer actively maintained. To find all solutions
  to this problem (and practice coding more problems) at:

  ~~~~~~~~~~~~~~~~~~~~~~~~~
  https://backtobackswe.com
  ~~~~~~~~~~~~~~~~~~~~~~~~~
*/

class Solution {
  public boolean isBalanced(TreeNode root) {
    return checkBalanced(root).isBalanced;
  }

  private BalanceStatusWithHeight checkBalanced(TreeNode root) {

    /*
     * Base case, an empty subtree is balanced and has a height of -1 as we define
     * it (since it is technically below "sea level", weird analogy)
     */
    if (root == null) {
      return new BalanceStatusWithHeight(true, -1);
    }

    /*
     * Go deep into the left subtree and get a result back
     */
    BalanceStatusWithHeight leftResult = checkBalanced(root.left);
    if (!leftResult.isBalanced) {
      return leftResult; // Left subtree is not balanced. Bubble up failure.
    }

    /*
     * Go deep into the right subtree and get a result back
     */
    BalanceStatusWithHeight rightResult = checkBalanced(root.right);
    if (!rightResult.isBalanced) {
      return rightResult; // Right subtree is not balanced. Bubble up failure.
    }

    /*
     * Our left and right subtrees are back and we have our results. Let us analyze
     * them here and bubble up our own answer.
     * 
     * 1.) Check if the subtreesAreBalanced 2.) Notate the height that this node
     * sits at (which is 1 plus the height of the larger of the left and right
     * subtrees coming off of this node)
     */
    boolean subtreesAreBalanced = Math.abs(leftResult.height - rightResult.height) <= 1;
    int height = Math.max(leftResult.height, rightResult.height) + 1;

    return new BalanceStatusWithHeight(subtreesAreBalanced, height);
  }

  /*
    We care about 2 things as our recursion goes upwards after bottoming out:
    1.)  The node's height
    2.)  Whether its left and right subtrees are balanced
  */
  private class BalanceStatusWithHeight {
    public boolean isBalanced;
    public int height;

    public BalanceStatusWithHeight(boolean isBalanced, int height) {
      this.isBalanced = isBalanced;
      this.height = height;
    }
  }
}

/* Solution via EPI (Elements of Programming Interviews) */
